src/hotspot/share/oops/oop.hpp

```diff
@@ -311,16 +342,44 @@ class oopDesc {
 
   // for code generation
   static int mark_offset_in_bytes()      { return (int)offset_of(oopDesc, _mark); }
-  static int klass_offset_in_bytes()     { return (int)offset_of(oopDesc, _metadata._klass); }
+  static int klass_offset_in_bytes()     {
+#ifdef _LP64
+    if (UseCompactObjectHeaders) {
+      // NOTE: The only place where this is used with compact headers is
+      // the C2 compiler, and even there we don't use it to access the (narrow)Klass*
+      // directly. It is used only as a placeholder to identify the special memory slice
+      // of LoadNKlass instructions. This value could be any value that is not a valid
+      // field offset. Use an offset halfway into the markWord, as the markWord is never
+      // partially loaded from C2.
+      return mark_offset_in_bytes() + sizeof(markWord) / 2;
+    } else
+#endif
```

Comment:
Maybe instead of trying to calculate some random, meaningless value just use some "random" value directly?
I am fine with the existing code, but first stating directly that "any value" works here, this additional code seems to confuse the message. (Fwiw, the method is also used during Universe initialization).

Reply:
Just to be clear, the second part of the quoted sentence is important:
> could be any value *that is not a valid field offset*

Reply:
>     could be any value that is not a valid field offset

I understand that that "random value" needs to satisfy this condition.

Reply:
With compact headers, this value should only be used in C2, and not really as an actual offset. An earlier version of the change had the value in src/hotspot/share/opto/type.hpp instead, and only an assert(!UCOH) in oopDesc::klass_offset_in_bytes(). I think this would be a better solution overall, because it prevents accidental (and wrong) usage of the klass_offset in the runtime. Back then it has been rejected by somebody (don't remember), because it made the C2 diff a little messier, so I kept it like it is now. I would prefer to reinstate it, though.

Reply:
> (Fwiw, the method is also used during Universe initialization).

Yes, but only in the -UCOH branch.

Reply:
We will deal with it in a follow-up. https://bugs.openjdk.org/browse/JDK-8340453